contrib/devtools/update-copyright-headers.py

```diff
@@ -0,0 +1,188 @@
+"""
+
+import argparse
+import os
+import re
+from datetime import datetime
+
+import git
+
+DEFAULT_COPYRIGHT_HOLDER = "The Horizen Foundation"
+DEFAULT_EXCLUDED_SUBDIRS = ["leveldb", "snark", "secp256k1", "univalue", "zcash"]
+DEFAULT_INCLUDED_EXTENSIONS = [".cpp", ".h", ".hpp"]
+
+# This is a safe date immediately before PR #613, which changed the header from Zen Blockchain Foundation to The Horizen Foundation
+DEFAULT_SINCE_DATE = datetime.fromisoformat("2023-09-28")
+DEFAULT_UNTIL_DATE = datetime.now()
+
+
+
+def parse_arguments():
+    """
+    Parse command-line arguments and return the parsed arguments.
+    """
+    parser = argparse.ArgumentParser(description="List files in a directory recursively")
+    parser.add_argument("directory", nargs='?', default="src",
+                        help="Path to the directory (default: 'src')")
+    parser.add_argument("--exclude-dir", nargs='?', default=DEFAULT_EXCLUDED_SUBDIRS,
+                        help=f"Subfolders to exclude from listing, by default {DEFAULT_EXCLUDED_SUBDIRS}")
+    parser.add_argument("--extensions", nargs='?', default=DEFAULT_INCLUDED_EXTENSIONS,
+                        help=f"Extensions of files to include in the list, by default {DEFAULT_INCLUDED_EXTENSIONS}")
+    parser.add_argument("--since", nargs='?', default=DEFAULT_SINCE_DATE, type=lambda s: datetime.strptime(s, '%Y-%m-%d'),
+                        help=f"Start date (format: YYYY-MM-DD) for filtering changed files, by default {DEFAULT_SINCE_DATE}")
+    parser.add_argument("--until", nargs='?', default=DEFAULT_UNTIL_DATE, type=lambda s: datetime.strptime(s, '%Y-%m-%d'),
+                        help="End date (format: YYYY-MM-DD) for filtering changed files, by default current time (now)")
+    parser.add_argument("--copyright-holder", default=DEFAULT_COPYRIGHT_HOLDER,
+                        help=f"Copyright holder name (default: '{DEFAULT_COPYRIGHT_HOLDER}')")
+    return parser.parse_args()
+
+def validate_arguments(args):
+    """
+    Validate the input arguments.
+    """
+    # Check if the provided path corresponds to a directory
+    if not os.path.isdir(args.directory):
+        print("Error: Invalid directory path.")
+        exit(1)
+
+    # Check if the specified excluded subfolders exist
+    for folder in args.exclude_dir:
+        path = os.path.join(args.directory, folder)
+        if not os.path.exists(path):
+            print(f"Error: Folder '{path}' does not exist in the specified directory.")
+            exit(1)
+
+    # Check if both --since and --until are provided
+    if (args.since and args.until and args.since > args.until):
+        print("Error: until_date must be higher than since_date.")
+        exit(1)
+
+def list_files(path, exclude_folders, extensions):
+    """
+    Recursively traverse a directory and return a list of files with specified extensions.
+    """
+    file_list = []
+    for root, dirs, files in os.walk(path):
+        # Exclude specified subfolders
```

Comment:
From this code I understand the `args.exclude_dir` set is targeting only the first level of `args.directory` (not deeper subdirectories):

```
    # Check if the specified excluded subfolders exist
    for folder in args.exclude_dir:
        path = os.path.join(args.directory, folder)
        if not os.path.exists(path):
            print(f"Error: Folder '{path}' does not exist in the specified directory.")
            exit(1)
```

if my understanding is correct, probably we want to modify a little bit `list_files` function:

```
    file_list = []
    first_level = True
    for root, dirs, files in os.walk(path):
        if first_level:
            # Exclude specified subfolders
            dirs[:] = [d for d in dirs if d not in exclude_folders]
            first_level = False
        for file in files:
```

in the situation we have deeper subdirectories named the same way (not the case as of now).

Reply:
Thanks for highlighting this, I improved the script by making the handling of paths more robust, combining relative and absolute paths.

This should now guarantee even more flexibility without the risk of multiple matches for the same folder.
In particular, supposing to have a folder tree like:
```
src/
├─ A/
├─ B/
├─ C/
│  ├─ A/
```
excluding the folder `A` wouldn't also cause `C/A` to be excluded, as it should be provided as a relative path to `src`.

Reply:
It's fine to me now, thanks.